src/eddymotion/model/dipy.py

```diff
@@ -0,0 +1,209 @@
+# emacs: -*- mode: python; py-indent-offset: 4; indent-tabs-mode: nil -*-
+# vi: set ft=python sts=4 ts=4 sw=4 et:
+#
+# Copyright 2024 The NiPreps Developers <nipreps@gmail.com>
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# We support and encourage derived works from this project, please read
+# about our expectations at
+#
+#     https://www.nipreps.org/community/licensing/
+#
+"""DIPY-like models (a sandbox to trial them out before upstreaming to DIPY)."""
+from __future__ import annotations
+
+import numpy as np
+
+from sklearn.gaussian_process import GaussianProcessRegressor
+from dipy.core.gradients import GradientTable
+from dipy.reconst.base import ReconstModel
+from dipy.reconst.multi_voxel import multi_voxel_fit
+
+
+def gp_prediction(
+    model: GaussianProcessRegressor,
+    gtab: np.ndarray,
+    mask: np.ndarray | None = None
+) -> tuple[np.ndarray, np.ndarray]:
+    """
+    Predicts one or more DWI orientations given a model.
+
+    This function checks if the model is fitted and then extracts
+    orientations and potentially b-values from the gtab. It predicts the mean
+    and standard deviation of the DWI signal using the model.
+
+    Parameters
+    ----------
+    model: :obj:`~sklearn.gaussian_process.GaussianProcessRegressor`
+        A fitted GaussianProcessRegressor model.
+    gtab: :obj:`~dipy.core.gradients.GradientTable`
+        A gradient table containing diffusion encoding information.
+    mask: :obj:`numpy.ndarray`
+        A boolean mask indicating which voxels to use (optional).
+
+    Returns
+    -------
+    :obj:`numpy.ndarray`
+        A 3D or 4D array with the simulated gradient(s).
+
+    """
+
+    # Check it's fitted as they do in sklearn internally
+    # https://github.com/scikit-learn/scikit-learn/blob/972e17fe1aa12d481b120ad4a3dc076bae736931/\
+    # sklearn/gaussian_process/_gpr.py#L410C9-L410C42
+    if not hasattr(model._gpr, "X_train_"):
+        raise RuntimeError("Model is not yet fitted.")
+
+    # Extract orientations from gtab, and highly likely, the b-value too.
+    return model._gpr.predict(gtab, return_std=False)
```

Comment:
@jhlegarreta I'm going to walk back my words and now I'm confident that the reshaping (w/ or w/o mask) should happen at a higher level of the class hierarchy. Basically, all models we have so far need to do it, so let's not re-implement the same operation for every model.

WDTY?

Reply:
Seems reasonable. Did not have time to have a look yesterday, so I am probably missing some context, but if you now have a clearer picture across all models and believe so, I am fine with it.